cpp/daal/src/data_management/finiteness_checker_cpu.cpp

```diff
@@ -0,0 +1,917 @@
+        }
+    });
+
+    st |= safeStat.detach();
+    if (!st)
+    {
+        return 0;
+    }
+
+    if (breakFlag)
+    {
+        sum         = getInf<double>();
+        sumIsFinite = false;
+    }
+    else
+    {
+        tlsSum.reduce([&](double * localSum) { sum += *localSum; });
+        tlsNotFinite.reduce([&](bool * localNotFinite) { sumIsFinite &= !*localNotFinite; });
+    }
+
+    return sum;
+}
+
+template <>
+double computeSumSOA<avx512>(NumericTable & table, bool & sumIsFinite, services::Status & st)
+{
+    return computeSumSOAAVX512Impl(table, sumIsFinite, st);
+}
+
+services::Status checkFinitenessInBlocks512(const float ** dataPtrs, bool inParallel, size_t nTotalBlocks, size_t nBlocksPerPtr, size_t nPerBlock,
+                                            size_t nSurplus, bool allowNaN, bool & finiteness)
+{
+    services::Status s;
+    const size_t nPerInstr = 16;
+    services::internal::TArray<bool, avx512> notFiniteArr(nTotalBlocks);
+    bool * notFinitePtr = notFiniteArr.get();
+    DAAL_CHECK_MALLOC(notFinitePtr);
+    for (size_t iBlock = 0; iBlock < nTotalBlocks; ++iBlock) notFinitePtr[iBlock] = false;
+
+    daal::conditional_threader_for(inParallel, nTotalBlocks, [&](size_t iBlock) {
+        size_t ptrIdx        = iBlock / nBlocksPerPtr;
+        size_t blockIdxInPtr = iBlock - nBlocksPerPtr * ptrIdx;
+        size_t start         = blockIdxInPtr * nPerBlock;
+        size_t end           = blockIdxInPtr == nBlocksPerPtr - 1 ? start + nPerBlock + nSurplus : start + nPerBlock;
+        size_t lcSize        = end - start;
+
+        // create masks for exponent and fraction parts of FP type and zero register
+        __m512i exp512Mask  = _mm512_set1_epi32(floatExpMask);
+        __m512i frac512Mask = _mm512_set1_epi32(floatFracMask);
+        __m512i zero512     = _mm512_setzero_si512();
+
+        __mmask16 notAllowNaNMask =
+            allowNaN ? _cvtu32_mask16(0) : _cvtu32_mask16(static_cast<unsigned int>(services::internal::MaxVal<int>::get()) * 2 + 1);
+        __mmask16 endMask = _cvtu32_mask16(0);
+
+        __m512i * ptr512i = (__m512i *)(dataPtrs[ptrIdx] + start);
+
+        for (size_t i = 0; i < lcSize / nPerInstr; ++i)
+        {
+            // apply masks
+            __m512i expBits  = _mm512_and_si512(exp512Mask, ptr512i[i]);
+            __m512i fracBits = _mm512_and_si512(frac512Mask, ptr512i[i]);
+
+            __mmask16 expAreOnes   = _mm512_cmpeq_epi32_mask(exp512Mask, expBits);
+            __mmask16 fracAreZeros = _mm512_cmpeq_epi32_mask(zero512, fracBits);
+
+            // "values aren't finite" = "exponent bits are ones" AND ( "fraction bits are zeros" OR NOT "NaN is allowed" )
+            __mmask16 orMask    = _kor_mask16(fracAreZeros, notAllowNaNMask);
+            __mmask16 finalMask = _kand_mask16(expAreOnes, orMask);
+            endMask             = _kor_mask16(endMask, finalMask);
+        }
+        if (_cvtmask16_u32(endMask) != 0) notFinitePtr[iBlock] = true;
```

Comment:
Note to reviewers: I have changed these lines to remove branching in the for loop. This was otherwise hidden due to the movement of the file, but otherwise the AVX512 portion is unchanged.